warp/backend.go

```diff
@@ -174,3 +197,39 @@ func (b *backend) GetMessage(messageID ids.ID) (*avalancheWarp.UnsignedMessage,
 
 	return unsignedMessage, nil
 }
+
+// addOffChainMessage parses the message and payload and adds to off-chain map if valid
+func (b *backend) addOffChainMessage(message hexutil.Bytes) error {
+	unsignedMessage, err := avalancheWarp.ParseUnsignedMessage(message)
+	if err != nil {
+		return fmt.Errorf("%w: %s", warp.ErrInvalidWarpMsg, err)
+	}
+
+	id, err := parsePayload(unsignedMessage)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Adding warp message to off-chain message map", "messageID", unsignedMessage.ID(), "decrypted warp message", hex.EncodeToString(unsignedMessage.Bytes()))
+	b.offChainMessageMap[id] = unsignedMessage
+	return nil
```

Comment:
Since there are two separate handlers for hash payloads vs. messages, we should create a separate map for the two different types. If a client mistakenly sends a request to the block hash handler with a messageID specified, we should not return a signature over a warp message containing an addressed call payload, since that is an unexpected response for that handler.

Reply:
got it